Poll consumer

macasado86 opened the pull request:
Hi Silviu,

This is the PR according to the [issue](https://github.com/silviucpp/erlkaf/issues/78) I opened about the explicit poll consumer. 

Instead of creating a `consumer_behaviour` to implement by anyone who wants to implement their own consumer, I have created a `consumer_group_behaviour` module to do that at consumer group level. The main reason is that I needed to create a `poll` function inside the consumer_group module and perhaps it is preferable to keep it separated from the current functionality. Anyway, I can change it if necessary

silviucpp commented:
Hello, 

I didn't reviewed yet but I see travis builds are broken: https://app.travis-ci.com/github/silviucpp/erlkaf/jobs/631024352

macasado86 commented:
Hi,

I have changed the second commit and now travis builds pass correctly.

Thanks

silviucpp commented:
Also documentation need to be updated https://github.com/silviucpp/erlkaf/blob/master/README.MD#consumer-api to include the second polling mechanism.

macasado86 commented:
Sorry, I forgot it. Now you have it.

macasado86 commented:
Hi Silviu,

Were you able to take a look at the PR?

Thanks

Miguel

silviucpp commented:
Hello Miguel! I've been incredibly busy lately, and I still am. However, I took a look, and everything seems fine.

My main concern is that erlkaf_poll_consumer_group and erlkaf_consumer_group share a lot of common logic, and I’m not sure what the most elegant way to handle this would be.

I haven't analyzed it in depth, but my initial plan was:

    Solution 1: If the changes aren’t too drastic, adapt erlkaf_consumer_group to work for both cases.

    Solution 2: If the code becomes too complex, keep erlkaf_consumer_group and create two separate modules with dedicated logic for each group type, while maintaining a shared interface that erlkaf_consumer_group would call.

The key issue is that logic like revoking/assigning partitions and closing the group is shared. If we introduce a bug in these areas, we'd need to fix it in multiple places.

Does it make sense for you ?

Silviu

macasado86 commented:
Hi,

I was thinking the same thing while doing the PR. I think all the logic regarding the assigning and revoking partitions should be only in a single place. Even the OAuthbearer refresh callback shouldn’t be repeated in multiple places.

I like both solutions and I think both can be valid. If I had to choose, maybe I would prefer the first one because the `poll_consumer_group` includes all the logic from the `consumer_group` plus two new functions: `poll` and `commit_offsets` (also the `valid_consumer_topics` function should have different implementations). So if we move all the common logic to a new module, maybe the `consumer_group` can become a bit ‘empty’.

Anyway, I think both approaches can fit. Which one do you prefer?

Miguel

silviucpp commented:
Let's go with solution 1. :)) for this reason I put it first . 

```Even the OAuthbearer refresh callback shouldn’t be repeated in multiple places.``` - I totally agree with you !

Give me a ping when you commit ! I receive lot of notification from github and I my miss notification from them

Silviu

macasado86 commented:
Hi Silviu,

I have changed the PR according to what to said yesterday. I have removed the `erlkaf_poll_consumer_group` module and
I have added the `poll` and `commit_offsets` functions to `erlkaf_consumer_group`.

Please let me know what you think and if there is anything else that needs to be modified.

Miguel

macasado86 commented:
Hello Silviu!

Yes, this last modification has been running in production for 24 hours. (The previous one, before our last conversation had been running for several days).

Miguel
